mdp/nodes/basis_function_nodes.py

```diff
@@ -0,0 +1,466 @@
+    returns an image of the basis function responses
+    for uniformly selected inputs within the given lims. This is
+    useful for visualizing the functions and debugging.
+
+    """
+
+    def __init__(self, basis_name, lims, order=3, decoupled=False, basis_params=None, scale_out=1.0, n_grid_pts=None,
+                 input_dim=None, dtype=None):
+        """
+        basis_name - name of the basis function to use: Supported: 'indicator', 'fourier', 'polynomial', 'radial',
+                                'lem' (laplacian eigen maps), 'gsfa' (graph-based slow features).
+        lims - a tuple of lower and upper bounds for each dimension of the input.
+                Eg., [(lower1, lower2, ...,), (upper1, upper2, ...,)]
+
+        order - order of the basis functions. A value that determines the dimensionality of the space
+                that the basis spans.
+                        Eg., order 4 of polynomial basis translates x to [1, x, x^2, x^3]
+                        Eg., order 30 of LEMs return top 30 laplacian eigen-maps excluding the constant function.
+                        Note that the first basis function (a constant function) for LEM and GSFA is filtered out.
+
+        decoupled - When True, basis functions do not include combinations of the input dims.
+                        Eg., order 3 of polynomial basis translates x to [1, x, x^2, y, y^2]
+                        If False, x -> [1, x, x^2, y, y^2, xy, xy^2, x^2y, x^2y^2].
+
+        basis_params - Additional (required or optional) params (list or a dict) that define the basis.
+
+        scale_out - scale output by a contant. y = basisfn(x)*scale_out
+
+        n_grid_pts - number of grid points within each dimension. A parameter used for discrete inputs.
+                     Eg. [n1, n2, ...,]
+
+        """
+        super(BasisFunctionNode, self).__init__(input_dim, output_dim=None, dtype=dtype)
+        self.basis_name = basis_name
+
+        if len(lims) != 2:
+            raise mdp.NodeException("'lims' has %d elements given, required 2 "
+                                    "[(lower1, lower2, ...,), (upper1, upper2, ...,)]" % (len(lims)))
+        if mdp.numx.isscalar(lims[0]):
+            lims = [tuple((lim,)) for lim in lims]
+        if len(lims[0]) != len(lims[1]):
+            raise mdp.NodeException("Length of lower_bounds ('lims[0]=%d') does not match the length "
+                                    "of the upper_bounds ('lims[1]=%d)." % (len(lims[0]), len(lims[1])))
+        self.lims = mdp.numx.asarray(lims)
+
+        self.order = order
+        self.scale_out = scale_out
+        self.decoupled = decoupled
+        self.basis_params = basis_params
+
+        self._gp = GridProcessingNode(grid_lims=lims, n_grid_pts=n_grid_pts)
+
+        self._input_dim = len(lims[0])
+
+        self._basis_mode = 'continuous'
+
+        # Basis Fns Init
+
+        if self.basis_name is 'indicator':
```

Comment:
This method is getting quite long ;-). You could either have subclasses for the different function types, or use something like the strategy pattern (which is probably nicer). If you use the strategy pattern then this might also be a chance to extract the get_fn_responses_img helper method, since I don't think that something like this should be part of a node class.